bakery/codec.go

```diff
@@ -230,28 +311,69 @@ func decodeCaveatIdV2(key *KeyPair, id []byte) (*ThirdPartyCaveatInfo, error) {
 		FirstPartyPublicKey: firstPartyPub,
 		ThirdPartyKeyPair:   *key,
 		RootKey:             rootKey,
-		CaveatId:            origId,
-		MacaroonId:          origId,
-		Version:             Version2,
+		Caveat:              origCaveat,
+		Version:             version,
+		Namespace:           ns,
 	}, nil
 }
 
-func decodeSecretPartV2(data []byte) (rootKey []byte, condition string, err error) {
+func decodeSecretPartV2V3(version Version, data []byte) (rootKey []byte, ns *checkers.Namespace, condition string, err error) {
 	if len(data) < 1 {
-		return nil, "", errgo.New("secret part too short")
+		return nil, nil, "", errgo.New("secret part too short")
 	}
-
-	version, data := data[0], data[1:]
-	if version != byte(Version2) {
-		return nil, "", errgo.Newf("unsupported secret part version %d", version)
+	gotVersion, data := data[0], data[1:]
+	if version != Version(gotVersion) {
+		return nil, nil, "", errgo.Newf("unexpected secret part version, got %d want %d", gotVersion, version)
 	}
 
 	l, n := binary.Uvarint(data)
 	if n <= 0 || uint64(n)+l > uint64(len(data)) {
-		return nil, "", errgo.Newf("invalid root key length")
+		return nil, nil, "", errgo.Newf("invalid root key length")
 	}
 	data = data[n:]
+	rootKey, data = data[:l], data[l:]
+
+	if version >= Version3 {
+		var nsData []byte
+		var ns1 checkers.Namespace
+
+		l, n = binary.Uvarint(data)
+		if n <= 0 || uint64(n)+l > uint64(len(data)) {
+			return nil, nil, "", errgo.Newf("invalid namespace length")
+		}
+		data = data[n:]
+		nsData, data = data[:l], data[l:]
+		if err := ns1.UnmarshalText(nsData); err != nil {
+			return nil, nil, "", errgo.Notef(err, "cannot unmarshal namespace")
+		}
+		ns = &ns1
+	} else {
+		ns = legacyNamespace()
+	}
+	return rootKey, ns, string(data), nil
+}
 
-	rootKey, condition = data[:l], string(data[l:])
-	return rootKey, condition, nil
+// appendUvarint appends n to data encoded as a variable-length
+// unsigned integer.
+func appendUvarint(data []byte, n uint64) []byte {
+	// Ensure the capacity is sufficient. If our space calculations when
+	// allocating data were correct, this should never happen,
+	// but be defensive just in case.
+	for need := uvarintLen(n); cap(data)-len(data) < need; {
+		data1 := append(data[0:cap(data)], 0)
+		data = data1[0:len(data)]
+	}
+	nlen := binary.PutUvarint(data[len(data):cap(data)], n)
+	return data[0 : len(data)+nlen]
+}
+
+// uvarintLen returns the number of bytes that n will require
+// when encoded with binary.PutUvarint.
+func uvarintLen(n uint64) int {
```

Comment:
hmm does this really pull its weight? are we so worried about memory that the extra bytes from using the MaxVarInt constants cannot be spared?